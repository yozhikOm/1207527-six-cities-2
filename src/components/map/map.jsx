import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import leaflet from 'leaflet';

class Map extends PureComponent {
  constructor(props) {
    super(props);

    this._map = null;
    this._zoom = 12;
    this._markersGroup = null;
  }

  componentDidMount() {
    const {currentCityCoords, offersArray, activeItemID} = this.props;

    this._map = this._initMap(this._zoom, currentCityCoords);
    this._markersGroup = leaflet.layerGroup().addTo(this._map);
    this._renderMap(this._map, this._zoom, currentCityCoords, offersArray, activeItemID);

  }

  componentDidUpdate() {
    const {currentCityCoords, offersArray, activeItemID} = this.props;
    this._markersGroup.clearLayers();
    this._renderMap(this._map, this._zoom, currentCityCoords, offersArray, activeItemID);
  }

  _initMap(zoom, currentCityCoords) {
    const map = leaflet.map(`map`, {
      center: currentCityCoords,
      zoom,
      zoomControl: false,
      marker: true
    });

    leaflet
      .tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
        attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
      })
      .addTo(map);

    return (map);
  }

  _renderMap(map, zoom, currentCityCoords, offersArray, activeItemID) {
    const icon = leaflet.icon({
      iconUrl: `/img/pin.svg`,
      iconSize: [30, 30]
    });
    const activeIcon = leaflet.icon({
      iconUrl: `/img/pin-active.svg`,
      iconSize: [30, 30]
    });

    map.setView(currentCityCoords, zoom);
    offersArray.forEach((offer) => {
      let currentIcon = offer.id === activeItemID ? activeIcon : icon;
      leaflet
          .marker(offer.coordinates, {icon: currentIcon})
          .addTo(this._markersGroup);
    });

  }

  render() {
    return (
      <div id="map" style={{width: `100%`, height: `100%`}} ></div>
    );
  }

}

Map.propTypes = {
  currentCityCoords: PropTypes.arrayOf(PropTypes.number.isRequired, PropTypes.number.isRequired).isRequired,
  offersArray: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    coordinates: PropTypes.arrayOf(PropTypes.number.isRequired, PropTypes.number.isRequired),
  })),
  activeItemID: PropTypes.number.isRequired,
};

Map.defaultProps = {
  currentCityCoords: [52.38333, 4.9],
  activeItemID: -1,
};

export {Map};
