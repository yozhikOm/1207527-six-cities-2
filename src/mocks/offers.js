export const offers = [
  {
    "id": 1,
    "title": `Cozy house amsterdam`,
    "type": `House`,
    "price": 650,
    "photos": [
      {
        "src": `img/apartment-01.jpg`
      },
      {
        "src": `img/amsterdam.jpg`
      },
    ],
    "coordinates": [52.3909553943508, 4.85309666406198],
  },
  {
    "id": 2,
    "title": `Cosy room in spacious apartment`,
    "type": `Room`,
    "price": 45,
    "photos": [
      {
        "src": `img/apartment-02.jpg`
      },
      {
        "src": `img/amsterdam.jpg`
      },
    ],
    "coordinates": [52.369553943508, 4.85309666406198],
  },
  {
    "id": 3,
    "title": `Elegant private room in stylish house`,
    "type": `Room`,
    "price": 128,
    "photos": [
      {
        "src": `img/room.jpg`
      },
      {
        "src": `img/amsterdam.jpg`
      },
    ],
    "coordinates": [52.3909553943508, 4.929309666406198],
  },
  {
    "id": 4,
    "title": `Huge Tinyhouse in the green, between sea and city`,
    "type": `House`,
    "price": 1520,
    "photos": [
      {
        "src": `img/apartment-03.jpg`
      },
      {
        "src": `img/amsterdam.jpg`
      },
    ],
    "coordinates": [52.3809553943508, 4.939309666406198],
  }
];
