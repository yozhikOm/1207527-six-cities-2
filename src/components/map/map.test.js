import React from 'react';
import {Map} from './map.jsx';
import ShallowRenderer from 'react-test-renderer/shallow';
import {offers} from '../../mocks/offers.js';

it(`Map компонент рендерится корректно`, () => {
  const mockCoorsArray = offers.map((offer) => offer.coordinates);

  const renderer = new ShallowRenderer();
  renderer.render(<Map coordinatesArray={mockCoorsArray} />);

  const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot();
});
