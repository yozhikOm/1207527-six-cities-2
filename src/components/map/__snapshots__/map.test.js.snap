// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map компонент рендерится корректно 1`] = `
<section
  className="cities__map map"
>
  <div
    id="map"
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  />
</section>
`;
